File: src/lib/google/hmac.ts
import * as gcp from '@pulumi/gcp';
import { all, Resource } from '@pulumi/pulumi';
import * as vault from '@pulumi/vault';

import { GoogleRepositoryProjectData } from '../../model/data/google';
import { StringMap } from '../../model/map';
import { writeToVault } from '../util/vault/secret';
import { vaultProvider } from '../vault';

/**
 * Creates IAM for a Google project.
 *
 * @param {GoogleRepositoryProjectData} project the Google project
 * @param {gcp.serviceaccount.Account} serviceAccount the service account for the project
 * @param {StringMap<gcp.Provider>} providers the providers for all projects
 * @param {StringMap<vault.Mount>} vaultStores the vault stores
 * @param {Resource[]} dependencies the Pulumi dependencies
 */
export const createHmacKey = (
  project: GoogleRepositoryProjectData,
  serviceAccount: gcp.serviceaccount.Account,
  providers: StringMap<gcp.Provider>,
  vaultStores: StringMap<vault.Mount>,
  dependencies: Resource[],
) => {
  const key = new gcp.storage.HmacKey(
    `gcp-hmac-${project.repository}-${project.name}`,
    {
      serviceAccountEmail: serviceAccount.email,
      project: project.name,
    },
    {
      provider: providers[project.name],
      dependsOn: dependencies,
    },
  );

  writeToVault(
    'google-cloud-storage',
    all([key.accessId, key.secret]).apply(([accessId, secret]) =>
      JSON.stringify({
        access_key_id: accessId,
        secret_access_key: secret,
      }),
    ),
    vaultProvider,
    vaultStores[project.repository],
  );
};

File: pkg/lib/google/hmac.go
package google

import (
	"encoding/json"
	"fmt"

	"github.com/muhlba91/github-infrastructure/pkg/lib/config"
	"github.com/muhlba91/github-infrastructure/pkg/model/google"
	"github.com/muhlba91/pulumi-shared-library/pkg/lib/vault/secret"
	"github.com/pulumi/pulumi-gcp/sdk/v9/go/gcp"
	"github.com/pulumi/pulumi-gcp/sdk/v9/go/gcp/serviceaccount"
	"github.com/pulumi/pulumi-gcp/sdk/v9/go/gcp/storage"
	"github.com/pulumi/pulumi-vault/sdk/v7/go/vault"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// createHMACKey creates an HMAC key for the specified Google Cloud project and stores it in Vault.
// ctx: Pulumi context for resource management.
// project: Google Cloud project details.
// serviceAccount: Service account associated with the HMAC key.
// vaultStore: Vault mount where the HMAC key will be stored.
// provider: GCP provider for resource creation.
func createHMACKey(ctx *pulumi.Context,
	project *google.RepositoryProject,
	serviceAccount *serviceaccount.Account,
	vaultStore *vault.Mount,
	provider *gcp.Provider,
) error {
	key, err := storage.NewHmacKey(
		ctx,
		fmt.Sprintf("gcp-hmac-%s-%s", *project.Repository, *project.Name),
		&storage.HmacKeyArgs{
			ServiceAccountEmail: serviceAccount.Email,
			Project:             pulumi.String(*project.Name),
		},
		pulumi.Provider(provider),
		pulumi.DependsOn([]pulumi.Resource{serviceAccount}),
	)
	if err != nil {
		return err
	}

	pulumi.All(vaultStore.Path, key.AccessId, key.Secret).ApplyT(func(all []any) error {
		path, _ := all[0].(string)
		accessID, _ := all[1].(string)
		secretKey, _ := all[2].(string)

		value, _ := json.Marshal(map[string]string{
			"access_key_id":     accessID,
			"secret_access_key": secretKey,
		})

		_, vErr := secret.Write(ctx, &secret.WriteArgs{
			Path:  path,
			Key:   "google-cloud-storage",
			Value: pulumi.String(value),
			PulumiOptions: []pulumi.ResourceOption{
				pulumi.Provider(config.VaultProvider),
			},
		})
		return vErr
	})

	return nil
}
